flips/20210916-nft-metadata.md

```diff
@@ -0,0 +1,338 @@
+```
+struct IPFSImage {
+    let url: String
+    let ipfsHash: String
+    let width: UInt32
+    let height: UInt32
+    let contentType: String
+}
+```
+
+It could then be retrieved from the NFT using the following:
+```
+import ArtNFT from 0x01010101
+   
+let artwork = nft.resolveView(Type<ArtNFT.IPFSImage>()) as? ArtNFT.IPFSImage
+```
+
+Another suggestion for media from FIND is:
+```
+pub struct Media {
+    pub let data: String
+    pub let contentType: String
+    pub let protocol: String
+
+    init(data:String, contentType: String, protocol: String) {
+        self.data=data
+        self.protocol=protocol
+        self.contentType=contentType
+    }
+}
+```
+Examples here are:
+
+An image on IPFS:
+data: QmRAQB6YaCyidP37UdDnjFY5vQuiBrcqdyoW1CuDgwxkD4
+contentType: image/jpeg
+protocol: ipfs
+
+An HTTP image:
+data: https://test.find.xyz/find.png
+contentType: image/jpeg
+protocol: http
+
+An on-chain image:
+data: data:image/png;base64,SOMEPNGDATAURI/wD/
+contentType: image/jpeg
+protocol: data
+
+The problem here is that the on-chain image can be quite large and you might not want to resolve those every time you fetch the media view. So what if we made a method on the struct that allowed you to retrieve the data/content instead?
+
+
+The same NFT author may also want expose other metadata about the artwork:
+```
+struct Artwork {
+    let name: String
+    let dateCreated: UInt64
+    let medium: String
+    let artistName: String
+}
+
+// ...
+
+let artwork = nft.resolveView(Type<ArtNFT.Artwork>()) as? ArtNFT.Artwork
+```
+
+If an art piece had more than one image then they could be retrieved using an array type as such:
+```
+import ArtNFT from 0x01010101
+   
+// ...
+   
+let images = nft.resolveView(Type<[ArtNFT.IPFSImage]>()) as? [ArtNFT.IPFSImage]
+```
+
+Or if a more structured approach is desired the NFT author may want to create a composite struct:
+```
+struct ArtworkDetail {
+    let artwork: Artwork
+    let primaryImage: IPFSImage
+    let secondaryImage: IPFSImage
+}
+```
+
+Because this proposal relies on the Cadence type system, adding new view types for exposing various
+types of metadata is trivial and very robust. A few examples that one might use as a view type are:
+
+```
+// an image
+pub struct IPFSImage {
+    pub let url: String
+    pub let ipfsHash: String
+    pub let width: UInt32
+    pub let height: UInt32
+    pub let contentType: String
+}
+
+// a video
+pub struct IPFSVideo {
+    pub let url: String
+    pub let ipfsHash: String
+    pub let width: UInt32
+    pub let height: UInt32
+    pub let codec: String
+    pub let container: String
+    pub let contentType: String
+}
+
+// a file stored on the blockchain
+pub struct Mime {
+    pub let type: String
+    pub let bytes: [UInt8]
+}
+
+// a location on the planet
+pub struct GeoPoint {
+    pub let lat: UFix64
+    pub let lng: UFix64
+    pub let projection: String
+}
+
+// a file hosted by a web server with a hash for verifying it's contents
```

Comment:
this `it's` should be `its`

Reply:
Can you make a suggestion for this and then i will apply it @rudi-cilibrasi ?

Reply:
it's fixed already ty

Reply:
yeah i fixed it :)